wasm/leb128/read.go

```diff
@@ -7,66 +7,146 @@
 package leb128
 
 import (
+	"errors"
 	"io"
 )
 
-// ReadVarUint32 reads a LEB128 encoded unsigned 32-bit integer from r, and
-// returns the integer value, and the error (if any).
-func ReadVarUint32(r io.Reader) (uint32, error) {
-	var (
-		b     = make([]byte, 1)
-		shift uint
-		res   uint32
-		err   error
-	)
+// readVarUint reads an unsigned integer of size n defined in https://webassembly.github.io/spec/core/binary/values.html#binary-int
+// readVarUint panics if n>64.
+func readVarUint(r io.Reader, n uint) (uint64, error) {
+	if n > 64 {
+		panic(errors.New("leb128: n must <= 64"))
+	}
+	p := make([]byte, 1)
+	var res uint64
+	var shift uint
 	for {
-		if _, err = io.ReadFull(r, b); err != nil {
-			return res, err
+		_, err := io.ReadFull(r, p)
+		if err != nil {
+			return 0, err
 		}
-
-		cur := uint32(b[0])
-		res |= (cur & 0x7f) << (shift)
-		if cur&0x80 == 0 {
+		b := uint64(p[0])
+		switch {
+		case b < 1<<7 && b < 1<<n:
+			res += (1 << shift) * b
 			return res, nil
+		case b >= 1<<7 && n > 7:
+			res += (1 << shift) * (b - 1<<7)
+			shift += 7
+			n -= 7
+		default:
+			return 0, errors.New("leb128: invalid uint")
 		}
-		shift += 7
 	}
 }
 
-// ReadVarint32 reads a LEB128 encoded signed 32-bit integer from r, and
-// returns the integer value, and the error (if any).
-func ReadVarint32(r io.Reader) (int32, error) {
-	n, err := ReadVarint64(r)
-	return int32(n), err
-}
+func readVarUintRecur(r io.Reader, n uint) (uint64, error) {
```

Comment:
please move this function to the `_test.go` file (so we don't use it unadvertantly).